### DISCUSSION TITLE
Multiple users getting images crossing?

### DISCUSSION THREAD

**Discussion by p2ochuck** [ImageMagick/ImageMagick#6383] [General]
Why is this happening or how can it be fixed?  Any help/assistance/insight would be greatly appreciated.

Our web application is experiencing something unusual.  2 users upload images at the same time which kicks off ImageMagick to do -identify and then -convert to create thumbnails of the images.  I see the magick.exe kickoff twice in the Task Manager, but **sometimes** the images get crossed.

What do I mean by get crossed?

User 1 uploads image A to create the thumbnails and User 2 uploads image B to create thumbnails at the same time.  The result is that User 1 gets thumbnails from image B and User 2 get thumbnails from image A.  It doesn't happen always, so it is intermittent.  All the data captured in the database is correct for the record it is just the images themselves that get crossed.

So, it seems like ImageMagick is putting information in a location that all instances of ImageMagick can access and the wrong instance is grabbing the wrong data.

**System Information**
Windows Server 2016
32G RAM
Processor:  Intel(R) Xeon(R) CPU E3-1271 v3 @ 3.60GHz, 3601 Mhz, 4 Core(s), 8 Logical Processor(s)

**Programming Platform and Language**
ColdFusion 2018 Enterprise using cfexecute to run command line magick.exe -identify and -convert against pdf images.

**ImageMagick Version**
ImageMagick-7.1.0-57-Q16-HDRI-x64-dll (standard policy.xml, no changes made from default)

**-identify code**
cfexecute(
	variable = "pdfData",
	name = "C:\Program Files\ImageMagick\magick.exe",
	arguments = 'identify -format "%[width]|%[height]|%[resolution.x]|%[colorspace]|%[xmp:CreatorTool]" "#arguments.ImagePath#[0]"',
	timeout = 1800
);

**-convert code**
cfexecute(
	name = "C:\Program Files\ImageMagick\magick.exe",
	arguments = 'convert -density 300 -units PixelsPerInch "#arguments.OriginalImage#[0]" "#arguments.TempImage#"',
	timeout = 1800
);

**Comment by snibgo** [comment]
I know nothing about ColdFusion or cfexecute.

> User 1 uploads image A to create the thumbnails ...

You don't show any code that creates thumbnails, or resizes any images.

> So, it seems like ImageMagick is putting information in a location that all instances of ImageMagick can access and the wrong instance is grabbing the wrong data.

No. If you run two instances of `magick`, they do not share memory.

Perhaps variables like `#arguments.TempImage#` have single instances but are used in multiple contexts when you have two uploads at the same time.

**Comment by p2ochuck** [comment]
Not knowing anything about ColdFusion doesn't matter.  I was just trying to give a complete picture of the server and programming language used.

Sorry missed the second part of the -convert code.  Here is the whole sequence of the code
1st - I want to make sure the image is at 300 dpi and do a convert of PDF (1st page only) to JPG.
cfexecute(
name = "C:\Program Files\ImageMagick\magick.exe",
arguments = 'convert -density 300 -units PixelsPerInch "#arguments.OriginalImage#[0]" "#arguments.TempImage#"',
timeout = 1800
);
2nd - Take the created JPG and resize it to a smaller JPG.
cfexecute(
name = "C:\Program Files\ImageMagick\magick.exe",
arguments = 'convert "#arguments.TempImage#" -resize 480 -background White "#admin480Front#"',
timeout = 1800
);

**Comment by snibgo** [comment]
Why are you using "convert"? That is provided for backwards compatibility, and gives IM v6 syntax. For the better v7 syntax, remove the "convert" words.

Are you doing anything between those `magick` commands? If not, then you can combine them into one command.

Do you need the temporary file? If you combine the two `magick` commands into one command, you don't need to write to a temporary file.

**Comment by p2ochuck** [comment]
I didn't realize I could remove "convert".  I've been using IM for a long time and the code has been in place since IMv6, but I'm by no means an expert with IM.  I will remove the convert and test for performance.  This application is all about performance and speed, while creating the best quality thumbnails possible.

There is no code between the two calls.

I did have them combined in to one call, but found by splitting them in to 2 was about 5 time faster processing and performance.  I know it doesn't make sense, but all performance testing proved it.  It might be that I didn't have the arguments in the optimal order.  What would you recommend?

I found that doing a straight conversion of a pdf to a jpg (the temp file) and then using the jpg to create the thumbnails was the best performance.

**Comment by snibgo** [comment]
> I did have them combined in to one call, ...

When they were combined, did you have the problem that "User 1 gets thumbnails from image B and User 2 get thumbnails from image A"?

If you didn't have the problem, perhaps the separate commands are interrupted by the User 2, and that user overwrites values of variables used by User 1.

**Reply by p2ochuck** [reply]
They have always been split apart for years and this just showed up as an issue recently.
